library/aria.c

```diff
@@ -0,0 +1,1022 @@
+/*
+ *  ARIA implementation
+ *
+ *  Copyright (C) 2006-2017, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+/*
+ * This implementation is based on the following standards:
+ * [1] http://210.104.33.10/ARIA/doc/ARIA-specification-e.pdf
+ * [2] https://tools.ietf.org/html/rfc5794
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_ARIA_C)
+
+#include "mbedtls/aria.h"
+
+#include <string.h>
+
+#if defined(MBEDTLS_SELF_TEST)
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdio.h>
+#define mbedtls_printf printf
+#endif /* MBEDTLS_PLATFORM_C */
+#endif /* MBEDTLS_SELF_TEST */
+
+#if !defined(MBEDTLS_ARIA_ALT)
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
+    volatile unsigned char *p = (unsigned char*)v; while( n-- ) *p++ = 0;
+}
+
+/*
+ * 32-bit integer manipulation macros (little endian)
+ */
+
+#ifndef GET_UINT32_LE
+#define GET_UINT32_LE(n,b,i)                    \
+{                                               \
+    (n) = ( (uint32_t) (b)[(i)    ]       )     \
+        | ( (uint32_t) (b)[(i) + 1] <<  8 )     \
+        | ( (uint32_t) (b)[(i) + 2] << 16 )     \
+        | ( (uint32_t) (b)[(i) + 3] << 24 );    \
+}
+#endif
+
+#ifndef PUT_UINT32_LE
+#define PUT_UINT32_LE(n,b,i)                                    \
+{                                                               \
+    (b)[(i)    ] = (unsigned char) ( ( (n)       ) & 0xFF );    \
+    (b)[(i) + 1] = (unsigned char) ( ( (n) >>  8 ) & 0xFF );    \
+    (b)[(i) + 2] = (unsigned char) ( ( (n) >> 16 ) & 0xFF );    \
+    (b)[(i) + 3] = (unsigned char) ( ( (n) >> 24 ) & 0xFF );    \
```

Comment:
Random thought: I wonder what the compiler actually outputs for something like this. For example, in ARMv7-M there is a REV instruction that does exactly this operation. I am curious to know whether its being used. The same for the rotation macros (there is an ROR instruction) used throughout the library...

Reply:
I'm don't think `PUT_INT32_LE()` needs `rev` on little-endian platforms: it's basically just a cast (and cast + rev on big-endian platforms). I just checked and GCC and Clang seem to produce a fairly straightforward sequence of shift, or, and ldrb/strb for this and `GET_INT32_LE()`. I think the reason the compiler can't make these a single ldr/str it has not guarantees on how the bytes buffer is aligned. (Whether it's worth special-casing the case where the buffer is aligned probably depends on usage patterns.)

For rotation (in this file, it's `ARIA_P2`) common compilers use the rotation instruction. Tested with GCC and Clang for x86, ARM-GCC, Clang, armcc5 and armcc6 for ARMv6-M (and ARMv8-A except armcc5).

For other byte rearrangements (`ARIA_P3` aka `rev` and `ARIA_P1` aka `rev16`) see [this commit](https://github.com/ARMmbed/mbedtls/pull/1410/commits/fb0e4f0d1a34bc7eca5b87169925c5c086db6c35) for x86 and [that one](https://github.com/ARMmbed/mbedtls/pull/1410/commits/377b2b624d5a5d894965236f0ae7fe8d09813a44) for arm.